### PR TITLE
Fix JSON map fields type changed to string

### LINKED ISSUE #54966: Geojson and maps attribute not working
### What is the bug or the crash?

Good evening,
I downloaded from uMap a geojson that has a definitive map attribute (maps, key format, value, dictionary in Python),
loading it in QGIS reads it correctly, but if I try to populate the defined map field with the expression:

map('key A','value A','key B','value B')

because it is so populated, QGIS does not give apparent problems;
when I reload the same geojson file, I notice that the attribute is no longer map type but string type and the file is no longer usable in uMap

### Steps to reproduce the issue

I attach a geojson, on this:

1. Load it into QGIS
2. popular attribute _umap_option2_ with the expression:
`map('color','White','fillColor','Lime')`
3. save and close QGIS;
4. Restart QGIS and load the newly edited geojson file
5. the _umap_options_ field has become a string

https://mega.nz/file/QZ5gGKJR#Jll-m7PzQT8ouIlSVSv0lWzZwWcwV3KZyGHNgKX3q64

## OSGeo4W 64b Win 10
Versione di QGIS
3.28.11-Firenze
Revisione codice QGIS
b02458a196
Versione Qt
5.15.3
Versione Python
3.9.5
Versione GDAL/OGR
3.7.2
Versione PROJ
9.3.0
Versione database del Registro EPSG
v10.094 (2023-08-08)
Versione GEOS
3.12.0-CAPI-1.18.0
Versione SQLite
3.41.1
Versione PDAL
2.5.5
Versione client PostgreSQL
unknown
Versione SpatiaLite
5.1.0
Versione QWT
6.1.6
Versione QScintilla2
2.13.4
Versione SO
Windows 10 Version 2009

Plugins Python attivi
DataPlotly
4.0.3
felt
1.1.0
kmltools
3.1.31
Label On Off
0.1
latlontools
3.6.15
layer2kmz
1.5.3
nominatim_locator_filter
0.2.4
OpenAndFilterAttributeTable
0.1
osminfo
0.7.1
processing_r
3.1.1
profiletool
4.2.6
qconsolidate
1.1.0
qgis-raster-attribute-table-plugin-bugfix-gh16
1.0
Qgis2threejs
2.7.1
qmapshaper
0.8.2
QuickOSM
2.2.3
quick_map_services
0.19.33
redLayer
2.2
SelectionFilter
1.0
soar
1.0.2
StyleLoadSave
1.0
valuetool
3.0.17
wbt_for_qgis
1.0.9
db_manager
0.1.20
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by elpaso**
Fix  #54966 (requires GDAL >= 3.7)

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1601,7 +1603,9 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags, QgsFeatureId
           QString stringValue;
 
           if ( OGR_Fld_GetSubType( fldDef ) == OFSTJSON )
-            stringValue = jsonStringValue( attrVal );
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrprovider.cpp:1604]: Should QgsOgrProvider::jsonStringValue be removed now? It looks unused after these changes, and if it's buggy then we don't want someone accidentally using it in future...

**Comment by nyalldawson** [comment]
What do the binary changes in the gpkg files relate to? Is that intentional?

**Comment by elpaso** [comment]
> What do the binary changes in the gpkg files relate to? Is that intentional?

No, I think it is the test which modifies the files, I'll check it.

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/providers/ogr/qgsogrprovider.cpp:1604]: I don't think it's buggy, it's just don't needed. But I agree it can be removed if not used.

Comment (elpaso) [reply src/core/providers/ogr/qgsogrprovider.cpp:1604]: I checked and it's still used, when the subtype is not JSON. I'll leave it.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**